lexer/stateful/stateful.go

```diff
@@ -249,30 +424,28 @@ type Lexer struct {
 func (l *Lexer) Next() (lexer.Token, error) { // nolint: golint
 	parent := l.stack[len(l.stack)-1]
 	rules := l.def.rules[parent.name]
-next:
+
 	for len(l.data) > 0 {
 		var (
 			rule  *compiledRule
 			match []int
+			err   error
 		)
-		for _, candidate := range rules {
-			// Special case "Return()".
-			if candidate.Rule == returnToParent {
-				l.stack = l.stack[:len(l.stack)-1]
-				parent = l.stack[len(l.stack)-1]
-				rules = l.def.rules[parent.name]
-				continue next
-			}
-			re, err := l.getPattern(candidate)
-			if err != nil {
-				return lexer.Token{}, participle.Wrapf(l.pos, err, "rule %q", candidate.Name)
-			}
-			match = re.FindSubmatchIndex(l.data)
-			if match != nil {
-				rule = &candidate // nolint: scopelint
-				break
+
+		if rule, match, err = rules.tryMatch(l); err != nil {
+			if rule != nil {
+				return lexer.Token{}, participle.Wrapf(l.pos, err, `rule %q`, rule.Name)
 			}
+			return lexer.Token{}, err // FIXME
```

Comment:
Same as above, plus this has no context. Fix what? Delete the FIXME if it's not a real issue.